numba/cuda/compiler.py

```diff
@@ -506,38 +498,59 @@ class _Kernel(serialize.ReduceMixin):
     CUDA Kernel specialized for a given set of argument types. When called, this
     object launches the kernel on the device.
     '''
-    def __init__(self, llvm_module, name, pretty_name, argtypes, call_helper,
-                 link=None, debug=False, fastmath=False, type_annotation=None,
-                 extensions=[], max_registers=None, opt=True):
+
+    @global_compiler_lock
+    def __init__(self, py_func, argtypes, link=None, debug=False, inline=False,
+                 fastmath=False, extensions=None, max_registers=None, opt=True):
         super().__init__()
-        # initialize CUfunction
+
+        self.py_func = py_func
+        self.argtypes = argtypes
+        self.debug = debug
+        self.extensions = extensions or []
+
+        cres = compile_cuda(self.py_func, types.void, self.argtypes,
+                            debug=self.debug,
+                            inline=inline)
+        fname = cres.fndesc.llvm_func_name
+        args = cres.signature.args
+        lib, kernel = cres.target_context.prepare_cuda_kernel(cres.library,
+                                                              fname,
+                                                              args,
+                                                              debug=self.debug)
+
         options = {
-            'debug': debug,
+            'debug': self.debug,
             'fastmath': fastmath,
             'opt': 3 if opt else 0
         }
 
+        llvm_module = str(lib._final_module)
+        pretty_name = cres.fndesc.qualname
+        ptx = CachedPTX(pretty_name, llvm_module, options=options)
```

Comment:
this originally was `str(llvm_module)`, should it still be?

Reply:
The `str()` is now in `llvm_module = str(lib._final_module)` just above.

Reply:
Maybe the name is now confusing - perhaps it should be `llvm_ir` instead of `llvm_module`? Would that be clearer as to what's going into `CachedPTX`?

Reply:
After some thought, I've decided that it is clearer as `llvm_ir` anyway, so I've made this change.